File: Matlab/offaxis_interference_hologram.m
%           h ——波长（mm）；          I ——数字全息图；
%           L ——全息图宽度（mm）；    z0——记录全息图的距离(mm)；
%% Image Read and preprocess
XRGB=imread('./images/pku.jpg');
X0=rgb2gray(XRGB);
% figure,imshow(X0,[]);
[M0,N0]=size(X0);

%% Set Basic parameters
N1=min(M0,N0);
N=1024;          %模拟形成的全息图取样数
X1=imresize(X0,N/4/N1);
[M1,N1]=size(X1);
X=zeros(N,N);
X(N/2-M1/2+1:N/2+M1/2,N/2-N1/2+1:N/2+N1/2)=X1(1:M1,1:N1);
h=0.632e-3;      %波长(mm), 可按需要修改
k=2*pi/h;
% CCD：Charge Coupled Device（电荷耦合元件），是一种将图像转换为电信号的半导体元件。由类似棋盘的格状排列的小像素 (pixel) 组成。
pix=0.00465     ;%CCD像素宽度(mm), 可按需要修改
L=N*pix;         %CCD宽度(mm)
z0=1000;         %衍射距离(mm), 可按需要修改
L0=h*N*z0/L;     %物平面宽度(mm)
Y=double(X);
%a=ones(N,N);
b=rand(N,N)*2*pi;
f=Y.*exp(1i.*b);  %叠加随机相位噪声,形成振幅正比于图像的初始场复振幅

figstr=strcat('初始物平面宽度=',num2str(L0),'mm');
figure(1),imshow(X,[]),colormap(gray); xlabel(figstr);title('物平面图像');

%% Fresnell 
%---------------菲涅耳衍射的S-FFT计算开始
n=1:N;
x=-L0/2+L0/N*(n-1);	 					
y=x;
[yy,xx] = meshgrid(y,x); %网格化数据平面
Fresnel=exp(1i*k/2/z0*(xx.^2+yy.^2));
f2=f.*Fresnel;
Uf=fft2(f2,N,N);
Uf=fftshift(Uf);
x=-L/2+L/N*(n-1);   %CCD宽度取样(mm) 					
y=x;
[yy,xx] = meshgrid(y,x); 
phase=exp(1i*k*z0)/(1i*h*z0)*exp(1i*k/2/z0*(xx.^2+yy.^2));%菲涅耳衍射积分前方的相位因子
Uf=Uf.*phase;
%---------------S-FFT计算结束
figstr=strcat('模拟CCD宽度=',num2str(L),'mm');
figure(2),imshow(abs(Uf),[]),colormap(gray); xlabel(figstr);title('到达CCD平面的物光振幅分布');

%% Reference Light
%---------------形成0-255灰度级的数字全息图
fex=N/L;
Qx=(4-2.5)*L0/8/z0;          %按照优化设计定义参考光方向余弦
Qy=Qx;
x=[-L/2:L/N:L/2-L/N];
y=x;
[X,Y]=meshgrid(x,y);
Ar=max(max(abs(Uf)));        %按物光场振幅最大值定义参考光振幅
Ur=Ar*exp(1i*k*(X.*Qx+Y.*Qy));%参考光复振幅

%% Interference
Uh=Ur+Uf;                    %物光与参考光干涉
Wh=Uh.*conj(Uh);             %干涉场强度
Imax=max(max(Wh));
I=uint8(Wh./Imax*255);      %形成0-255灰度级的数字全息图
Ih=I;
imwrite(I,'./Ih.bmp');     %形成数字全息图文件
figstr=strcat('全息图宽度=',num2str(L),'mm');
figure(3),imshow(Ih,[]),colormap(gray);xlabel(figstr);title('模拟形成的数字全息图');

%% Reconstruction
%基于S-FFT的全息重建
f0 = double(Ih);
[N1,N2]=size(f0);
N=min(N1,N2);                      
h=0.000632;             %波长(mm)              
z0=1000;
L=N*pix;                %CCD宽度(mm)                        
In(1:N,1:N)=f0(1:N,1:N);
%-----------------------------1-FFT重建开始
n=1:N;
x=-L/2+L/N*(n-1);	 					
y=x;
[yy,xx] = meshgrid(y,x); 
k=2*pi/h;	
Fresnel=exp(1i*k/2/z0*(xx.^2+yy.^2));
f2=In.*Fresnel;
Uf=fft2(f2,N,N);
Uf=fftshift(Uf);
L0=h*z0*N/L;
x=-L0/2+L0/N*(n-1);	 					
y=x;
[yy,xx] = meshgrid(y,x); 
phase=exp(1i*k*z0)/(1i*h*z0)*exp(1i*k/2/z0*(xx.^2+yy.^2));
U0=Uf.*phase;   %积分运算结果乘积分号前方相位因子
%-----------------------------1-FFT重建结束
If=U0*conj(U0);
Gmax=max(max(abs(U0)));
Gmin=min(min(abs(U0)));
figstr=strcat('重建物平面宽度=',num2str(L0),'mm');
figure(4),
imshow(abs(U0),[Gmin Gmax/1]),colormap(gray); xlabel(figstr);title('1-FFT物平面重建图像');

%% Light intensity change
p=10;
while p
    figure(5);
    imshow(abs(U0),[Gmin Gmax/p]),colormap(gray);xlabel(figstr);title('1-FFT物平面重建图像');
    p=input('Gmax/p,p=10?');
end
Up=fft2(U0);
figure(6),imshow(abs(Up),[])
figure(7),plot(abs(Up(round(512)+1,:)));
hold on;

File: Python/offaxis_interference_hologram.py
import cv2
import numpy as np
from scipy.fftpack import fft2, ifft2, fftshift
import matplotlib.pyplot as plt

# input the image path
image_path = "../Res/imageO/pku.jpg"
img = cv2.imread(image_path)
img_gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
print(img_gray.shape)
p, q = img_gray.shape
# show the raw image
plt.figure(0)
plt.imshow(img_gray, cmap="gray")
plt.title("Raw Image")

# Set Basic parameters
N1 = min(p, q)
N = 1024    # 采样率
scale = 1/4

size_scale = N / N1 * scale
size_scale_x = int(size_scale * p)
size_scale_y = int(size_scale * q)
print(size_scale_x)

X1 = cv2.resize(img_gray, [size_scale_x, size_scale_y])

M1, N1 = X1.shape
X = np.zeros((N, N))
X[int(N/2 - M1/2 + 1):int(N/2 + M1/2), int(N/2 - N1/2 + 1):int(N/2 + N1/2)] = X1[1:M1, 1:N1]

h = 0.632e-3        # 波长，单位mm
k = 2 * np.pi / h
pix = 0.00465        # CCD像素宽度
L = N * pix         # CCD宽度
z0 = 1000           # 衍射距离
L0 = h * z0 / pix   # 重建像平面宽度

Y = X
b = np.random.rand(N, N) * 2 * np.pi
f = Y * np.exp(1j * b)     # 叠加随机相位噪声，形成振幅正比于图像的初始场复振幅
X0 = np.abs(f)

plt.figure(1)
plt.imshow(X, cmap="gray")
plt.title("Scale Image")

# Fresnell
n = np.array(range(N))
x = -L0 / 2 + L0 / N * n
y = x
yy, xx = np.meshgrid(y, x)
Fresnel = np.exp(1j * k / 2 / z0 * (xx * xx + yy * yy))
f2 = f * Fresnel
Uf = fft2(f2, (N, N))
Uf = fftshift(Uf)
x = -L / 2 + L / N * n
y = x
yy, xx = np.meshgrid(y, x)
phase = np.exp(1j * k * z0) / (1j * h * z0) * np.exp(1j * k / 2 / z0 * (np.power(xx, 2) + np.power(yy, 2)))
Uf = Uf * phase

plt.figure(2)
plt.imshow(np.abs(Uf), cmap="gray")
plt.title("Amplitude distribution of object light")

# Reference Light
Qx = (4 - 2.5) * L0 / 8 / z0
Qy = Qx
x = np.linspace(-L/2, L/2 - L/N, N)
y = x
X, Y = np.meshgrid(x, y)
Ar = np.max(np.abs(Uf))
Ur = Ar * np.exp(1j * k * (X * Qx + Y * Qy))

# Interference
Uh = Ur + Uf
Wh = Uh * np.conj(Uf)
Wh = np.abs(Wh)
Imax = np.max(Wh)
Ih = Wh / Imax * 255

plt.figure(3)
plt.imshow(Ih, cmap="gray")
plt.title("Interference Hologram")
cv2.imwrite("./result/oaih_pku_CGH.bmp", Ih)

# Reconstruction
N1, N2 = Ih.shape
N = min(N1, N2)
h = 0.000632        # 波长(mm)
z0 = 1000
L = N * pix         # CCD宽度(mm)
In = Ih

n = np.array(range(N))
x = -L / 2 + L / N * n
y = x
yy, xx = np.meshgrid(y, x)
k = 2 * np.pi / h

Fresnel = np.exp(-1j * k / 2 / z0 * (np.power(xx, 2) + np.power(yy, 2)))
f2 = In * Fresnel
Uf = ifft2(f2, (N, N))
Uf = fftshift(Uf)
L0 = h * z0 / pix

x = -L0 / 2 + L0 / N * n
y = x
yy, xx = np.meshgrid(y, x)
phase = np.exp(-1j * k * z0) / (-1j * h * z0) * np.exp(-1j * k / 2 / z0 * (np.power(xx, 2) + np.power(yy, 2)))
U0 = Uf * phase
U0 = abs(U0)
Gmax = np.max(U0)
Gmin = np.min(U0)
# U0 = U0 / Gmax * 255

# try to change the parameter to get the best image
p = 10
Gm = Gmax / p
np.clip(U0, Gmin, Gm, out=U0)

U = U0 / Gm * 255

plt.figure(4)
plt.imshow(U0, cmap="gray")
plt.title("Amplitude distribution of object plane reconstructed by inverse operation")
cv2.imwrite("./result/oaih_pku_recover.bmp", U)

plt.show()









